File: problems/DSA1/sorting/1.CostToRemoveElements.java
/**
 * Run:
 * & "C:\Program Files\Java\jdk-25.0.2\bin\javac.exe" problems/DSA1/sorting/1.CostToRemoveElements.java
 * & "C:\Program Files\Java\jdk-25.0.2\bin\java.exe" problems.DSA1.sorting.CostToRemoveElements
 * 
 * NOTE: Java is a compiled language. 
 * - When you modify this .java file, you MUST re-run 'javac' to update the .class file (bytecode).
 * - Running the code executes the .class file, not the .java file directly.
 */
package problems.DSA1.sorting;

import java.util.Arrays;

/**
 * Problem: Cost to Remove Elements
 * Difficulty: Easy
 * Category: Sorting
 * Description: Given an integer array A, find min cost to remove all elements. Cost = Sum of current elements.
 */
class CostToRemoveElements {

    /**
     * Calculates the minimum cost to remove all elements.
     * 
     * Time Complexity: O(N log N)
     * Space Complexity: O(1) or O(log N) for sort stack
     */
    public static int solve(int[] A) {
        if (A == null || A.length == 0) {
            return 0;
        }

        // Sort ascending: [Smallest, ..., Largest]
        // Time: O(N log N) | Space: O(log N) (Dual-Pivot Quicksort)
        Arrays.sort(A); // Time: O(N log N) | Space: O(log N)
        
        int n = A.length;
        int totalCost = 0;
        
        // A[0] is smallest (removed last, coeff N)
        // A[n-1] is largest (removed first, coeff 1)
        for (int i = 0; i < n; i++) {
            // Coeff = n - i
            totalCost += A[i] * (n - i);
        }

        return totalCost;
    }

    public static void main(String[] args) {
        // Dummy Data & Verification
        int[] input1 = {2, 1};
        int expected1 = 4;
        int result1 = solve(input1);
        System.out.println("Test 1 (Basic): " + (result1 == expected1 ? "PASS" : "FAIL") + " | Expected: " + expected1 + ", Got: " + result1);

        int[] input2 = {5};
        int expected2 = 5;
        int result2 = solve(input2);
        System.out.println("Test 2 (Single): " + (result2 == expected2 ? "PASS" : "FAIL") + " | Expected: " + expected2 + ", Got: " + result2);

        int[] input3 = {1, 2, 3};
        int expected2_custom = 10;
        int result3 = solve(input3);
        System.out.println("Test 3 (Three): " + (result3 == expected2_custom ? "PASS" : "FAIL") + " | Expected: " + expected2_custom + ", Got: " + result3);
    }
}

/*
 * Educational Summary:
 * 1. Arrays.sort(int[]): Uses Dual-Pivot Quicksort. Time Complexity is O(N log N). Space is O(log N).
 * 2. Greedy Approach: To minimize cost, we multiply larger numbers by smaller coefficients.
 *    Sorting ascending allows us to remove smaller elements last (highest coefficients).
 */

File: problems/DSA1/sorting/1.cost-to-remove-elements.js
/**
 * Run this problem:
 * node problems/DSA1/sorting/cost-to-remove-elements.js
 */

/**
 * Problem: Cost to Remove Elements
 * Difficulty: Easy
 * Category: Sorting
 * Source: Custom / Common Interview Question
 * 
 * Description:
 * Given an integer array A of size N. You can remove any element from the array in one operation.
 * The cost of this operation is the sum of all elements in the array present before this operation.
 * Find the minimum cost to remove all elements from the array.
 * 
 * Constraints:
 * - 0 <= N <= 1000
 * - 1 <= A[i] <= 1000
 * 
 * Examples:
 * Input: A = [2, 1]
 * Output: 4
 * Explanation:
 * 1. Remove 2: Cost = 2 + 1 = 3. Array becomes [1].
 * 2. Remove 1: Cost = 1. Array becomes [].
 * Total Cost = 3 + 1 = 4.
 * 
 * Input: A = [5]
 * Output: 5
 */

// ============================================
// SOLUTION 1: Greedy Approach (Sort Descending)
// ============================================
/**
 * Approach:
 * To minimize the total cost, we want to remove the largest elements when the array size is smallest,
 * effectively multiplying them by a smaller factor. However, looking at the cost structure:
 * Cost = Sum(Elements in iteration 1) + Sum(Elements in iteration 2) + ...
 * 
 * An element removed at step `k` (1-indexed) contributes to the sum `k` times.
 * Total Cost = Sum(A[i] * (position_of_removal_from_start))
 * 
 * To minimize this sum, we should assign smaller multipliers to larger values.
 * Wait, let's re-verify.
 * 
 * Example: [A, B]
 * Order A, then B:
 * Cost 1: A+B
 * Cost 2: B
 * Total: A + 2B
 * 
 * Order B, then A:
 * Cost 1: A+B
 * Cost 2: A
 * Total: 2A + B
 * 
 * If A > B, then A + 2B < 2A + B (subtract A+B from both sides -> B < A).
 * So if A > B, we want A to be removed first (multiplier 1) and B second (multiplier 2).
 * 
 * Generalizing: We should remove elements in Descending Order.
 * The largest element is removed 1st (multiplied by 1).
 * The 2nd largest is removed 2nd (multiplied by 2).
 * ...
 * The smallest element is removed last (multiplied by N).
 * 
 * Algorithm:
 * 1. Sort the array in descending order.
 * 2. Iterate through the sorted array.
 * 3. Add A[i] * (i + 1) to the total cost.
 * 
 * Time Complexity: O(N log N)
 * - Due to sorting.
 * 
 * Space Complexity: O(1)
 * - If we ignore the space taken by sort (or O(N) depending on engine).
 */
function solution1(A) {
    // Handle edge case for empty array
    if (!A || A.length === 0) return 0;

    // Sort in descending order
    // Note: copy array to avoid mutating input if that's a concern, though usually fine in CP
    // A.sort((a, b) => b - a); 
    // Creating a copy to be safe and pure
    const sortedA = [...A].sort((a, b) => b - a);

    let totalCost = 0;
    for (let i = 0; i < sortedA.length; i++) {
        totalCost += sortedA[i] * (i + 1);
    }

    return totalCost;
}

// ============================================
// RUN TESTS
// ============================================
// ============================================
// RUN TESTS
// ============================================
const { runTests } = require('../../../utils/test-runner');
const { testCases, validator } = require('../../../utils/DSA1/sorting/cost-to-remove-elements');

if (require.main === module) {
    // 1. Run Automated Test Runner
    runTests({
        problemName: "Cost to Remove Elements",
        solution: solution1,
        testCases: testCases,
        validator: validator,
        currentComplexity: { time: 'O(N log N)', space: 'O(1)' },
        optimalComplexity: { time: 'O(N log N)', space: 'O(1)' }
    });

    // 2. Dummy Data Execution Block (Visual Check)
    console.log("\n--- Standalone Execution ---");
    const dummyInput = [2, 1];
    console.log("Input:", dummyInput);
    console.log("Output:", solution1(dummyInput));
}

module.exports = { solution1 };

/*
 * Educational Summary:
 * - Array.prototype.sort(): In V8 (Chrome/Node), this typically uses TimSort. 
 *   Time Complexity: O(N log N). Space Complexity: O(N) due to creating a new array [...A] to avoid mutation.
 *   Note: In-place sort would be O(1) space but modifies input.
 */
